### PR TITLE
Repository avatars

### PR DESCRIPTION
**Pull request by sergey-dryabzhinsky**
- first variant of code from old work for gogs
- add migration 87
- add new option in app.ini
- add en-US locale string
- add new class in repository.less

For issue #694 and #5625

### CODE REVIEW

**Comment by sergey-dryabzhinsky** [comment]
Should I commit also index.css?

**Comment by lunny** [comment]
@sergey-dryabzhinsky yes, please.

**Comment by lunny** [comment]
add new option in app.ini.sample and docs

**Comment by sergey-dryabzhinsky** [comment]
Ready for review

File: routers/repo/setting.go [hunk 1]
```diff
@@ -727,3 +730,53 @@ func init() {
 		panic(err)
 	}
 }
+
+// UpdateAvatarSetting update repo's avatar
+// FIXME: limit size.
```
Comment (zeripath) [inline routers/repo/setting.go]: You must limit size before we merge - just stick a config in it doesn't need a great UI.

File: models/migrations/v87.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+package migrations
+
+import (
+	"fmt"
+
+	"github.com/go-xorm/xorm"
+)
+
+func addAvatarFieldToRepository(x *xorm.Engine) (err error) {
```
[the remark this answers is not in the document]
Comment (sergey-dryabzhinsky) [reply models/migrations/v87.go]: I just following interface for `NewMigration` from `models/migrations/migrations.go`

File: models/migrations/v87.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	case "tidb":
+		_, err = sess.Exec("ALTER TABLE repository ADD `avatar` TEXT")
+	case "mssql":
+		_, err = sess.Exec("ALTER TABLE repository ADD \"avatar\" VARCHAR")
+	case "sqlite3":
+		_, err = sess.Exec("ALTER TABLE repository ADD COLUMN `avatar` TEXT")
+	}
+
+	if err != nil {
+		return fmt.Errorf("Error changing mirror interval column type: %v", err)
```
[the remark this answers is not in the document]
Comment (sergey-dryabzhinsky) [reply models/migrations/v87.go]: Fixed

File: routers/repo/setting.go [hunk 4]
```diff
@@ -727,3 +730,53 @@ func init() {
...
+func UpdateAvatarSetting(ctx *context.Context, form auth.AvatarForm) error {
+	ctxRepo := ctx.Repo.Repository
+	if form.Avatar != nil {
+		r, err := form.Avatar.Open()
+		if err != nil {
+			return fmt.Errorf("Avatar.Open: %v", err)
+		}
+		defer r.Close()
+
+		data, err := ioutil.ReadAll(r)
```
[the remark this answers is not in the document]
Comment (sergey-dryabzhinsky) [reply routers/repo/setting.go]: I added File size check

File: routers/repo/setting.go (hunk 1 above)
Comment (sergey-dryabzhinsky) [reply routers/repo/setting.go]: Fixed

**Comment by sergey-dryabzhinsky** [comment]
Drone service fails build.
Seems like main repo for https://gitlab.com/strk/go-libravatar dead by now.

**Comment by NotAFile** [comment]
The avatars seem a bit large to me, any other opinons?

**Comment by sergey-dryabzhinsky** [comment]
@NotAFile in header on a repository page?

**Comment by sergey-dryabzhinsky** [comment]
Cleaner repository header template:
Снимок экрана_2019-05-19_16-24-10

File: models/migrations/v87.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	defer sess.Close()
+
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	dialect := x.Dialect().DriverName()
+
+	switch dialect {
+	case "mysql":
```
Comment (techknowlogick) [inline models/migrations/v87.go]: I need to use case, Xorm can handle this. You can use Sync2 method

**Review by techknowlogick — DISMISSED** [review]

File: models/migrations/v87.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sergey-dryabzhinsky) [reply models/migrations/v87.go]: Done

File: models/migrations/v87.go (hunk 5 above)
Comment (sergey-dryabzhinsky) [reply models/migrations/v87.go]: Done

File: models/migrations/v87.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lunny) [reply models/migrations/v87.go]: @zeripath @techknowlogick you are right. `Sync2` will be simpler.

File: models/repo.go [hunk 6]
```diff
@@ -2452,3 +2473,114 @@ func (repo *Repository) GetUserFork(userID int64) (*Repository, error) {
...
+}
+
+// UploadAvatar saves custom avatar for repository.
+// FIXME: split uploads to different subdirs in case we have massive number of repos.
+func (repo *Repository) UploadAvatar(data []byte) error {
+	imgCfg, _, err := image.DecodeConfig(bytes.NewReader(data))
+	if err != nil {
+		return fmt.Errorf("DecodeConfig: %v", err)
+	}
+	if imgCfg.Width > setting.AvatarMaxWidth {
```
Comment (lafriks) [inline models/repo.go]: Please reuse same code as recently added in user avatar upload

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/repo.go (hunk 6 above)
Comment (sergey-dryabzhinsky) [reply models/repo.go]: I updated to latest master. That code not changed.
Снимок экрана_2019-05-24_21-30-39

Can you point me to the desired code?

Comment (lafriks) [reply models/repo.go]: Oh, sorry I thought I merged #7025 already. It still needs @lunny approval

Comment (lafriks) [reply models/repo.go]: Now it's merged

Comment (sergey-dryabzhinsky) [reply models/repo.go]: Done

**Comment by sergey-dryabzhinsky** [comment]
Any other suggestions?

File: custom/conf/app.ini.sample [hunk 7]
```diff
@@ -504,6 +504,7 @@ SESSION_LIFE_TIME = 86400
 
 [picture]
 AVATAR_UPLOAD_PATH = data/avatars
+REPOSITORY_AVATAR_UPLOAD_PATH = data/repo-avatars
```
Comment (lunny) [inline custom/conf/app.ini.sample:507]: AVATAR_MAX_FILE_SIZE

**Review by lunny — CHANGES_REQUESTED** [review]

File: custom/conf/app.ini.sample (hunk 7 above)
Comment (sergey-dryabzhinsky) [reply custom/conf/app.ini.sample:507]: Missed that option.
Fixed in https://github.com/go-gitea/gitea/pull/6986/commits/c3f386cd42c17a50895b868836c6391839b12d41.

**Comment by sergey-dryabzhinsky** [comment]
@lunny @lafriks @techknowlogick Any resolutions on reviews?

**Comment by techknowlogick** [comment]
I've just dismissed my review so I'm not blocking this PR anymore.

File: models/repo.go [hunk 8]
```diff
@@ -166,6 +172,9 @@ type Repository struct {
 	CloseIssuesViaCommitInAnyBranch bool               `xorm:"NOT NULL DEFAULT false"`
 	Topics                          []string           `xorm:"TEXT JSON"`
 
+	// Avatar
+	Avatar string `xorm:"VARCHAR(2048)"`
```
Comment (techknowlogick) [inline models/repo.go]: Does the full URL need to be stored in DB? For user avatars this makes sense due to gravatar (and other external avatar services), but for repo avatars it follows the following patter `setting.AppSubURL + "/repo-avatars/" + repo.Avatar`

Comment (sergey-dryabzhinsky) [reply models/repo.go]: You mean that it easily fits into 250 symbols?
ID+MD5 ~ (10-20) + 32.
64 symbols must be enough I think.

Comment (techknowlogick) [reply models/repo.go]: Yes, rather than storing the full URL, just the filename could be stored.

Comment (sergey-dryabzhinsky) [reply models/repo.go]: Done in https://github.com/go-gitea/gitea/pull/6986/commits/4ae938309774a9d2e325fbbfcc36278294d728e6

File: models/repo.go [hunk 9]
```diff
@@ -2452,3 +2471,101 @@ func (repo *Repository) GetUserFork(userID int64) (*Repository, error) {
...
+	if err = sess.Begin(); err != nil {
+		return err
+	}
+
+	oldAvatarPath := repo.CustomAvatarPath()
+
+	// Users can upload the same image to other repo - prefix it with ID
+	// Then repo will be removed - only it avatar file will be removed
+	repo.Avatar = fmt.Sprintf("%d-%x", repo.ID, md5.Sum(data))
+	if _, err := x.ID(repo.ID).Cols("avatar").Update(repo); err != nil {
```
Comment (lunny) [inline models/repo.go]: `sess.` not `x.`

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/repo.go (hunk 9 above)
Comment (sergey-dryabzhinsky) [reply models/repo.go]: Done in https://github.com/go-gitea/gitea/pull/6986/commits/c007a9a947343f4a11589177e3872f51a709989d

**Comment by sergey-dryabzhinsky** [comment]
Also rewrite `models.repo.DeleteAvatar` - now it uses Session too.

File: models/repo.go [hunk 10]
```diff
@@ -2452,3 +2471,119 @@ func (repo *Repository) GetUserFork(userID int64) (*Repository, error) {
...
+	}
+
+	return sess.Commit()
+}
+
+// DeleteAvatar deletes the repos's custom avatar.
+func (repo *Repository) DeleteAvatar() error {
+
+	// Avatar exists
+	if len(repo.Avatar) > 0 {
```
Comment (lafriks) [inline models/repo.go]: Please return as soon as possible for code readability:
```suggestion
	if len(repo.Avatar) == 0 {
		return nil
	}
```

File: routers/repo/setting.go [hunk 11]
```diff
@@ -727,3 +730,56 @@ func init() {
 		panic(err)
 	}
 }
+
+// UpdateAvatarSetting update repo's avatar
+func UpdateAvatarSetting(ctx *context.Context, form auth.AvatarForm) error {
+	ctxRepo := ctx.Repo.Repository
+	if form.Avatar != nil {
```
Comment (lafriks) [inline routers/repo/setting.go]: Same here (do not use else):
```suggestion
	if form.Avatar == nil {
		// No avatar is uploaded but setting has been changed to enable
		// No random avatar here.
		if !com.IsFile(ctxRepo.CustomAvatarPath()) {
			log.Trace("No avatar was uploaded for repo: %d. Default icon will appear instead.", ctxRepo.ID)
		}
		return nil
	}
```

**Review by lafriks — CHANGES_REQUESTED** [review]
Two code readability issues but otherwise LG-TM

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
@sergey-dryabzhinsky Thanks for your good job!!!

**Comment by richmahn** [comment]
Yes, this looks great! Was going to review but looks like you have reviewers, and a few minor details to fix. This is definitely something my organization will find useful!

File: models/repo.go (hunk 10 above)
Comment (sergey-dryabzhinsky) [reply models/repo.go]: Point taken.
Done in https://github.com/go-gitea/gitea/pull/6986/commits/f53758af4662b3d462677d92fb2b1ed472ce7998

File: routers/repo/setting.go (hunk 11 above)
Comment (sergey-dryabzhinsky) [reply routers/repo/setting.go]: Done in https://github.com/go-gitea/gitea/pull/6986/commits/f53758af4662b3d462677d92fb2b1ed472ce7998

**Comment by sergey-dryabzhinsky** [comment]
@richmahn same here.
I've tired to make war with bloated GitLab in my company.
So I decided to take an action.

Love to use gitea at home to manage my pet projects.
Now it will be even better.

**Review by lafriks — APPROVED** [review]

**Comment by saitho** [comment]
When no repository avatar was assigned it will render an empty img tag in repository list.
``

**Comment by maznu** [comment]
Really hyped for 1.9! Thanks, everyone! 👍

**Comment by saitho** [comment]
> Really hyped for 1.9!

Definitely! To me 1.9 seems to be much larger than previous feature releases. 👍

**Pull request merged**